src/ethereum/EthereumEngine.ts

```diff
@@ -976,10 +979,40 @@ export class EthereumEngine extends CurrencyEngine<
       }
 
       // Update total fee:
-      const totalFee = add(primaryNetworkFee, rollupFee)
+      const totalFeeInitial = add(primaryNetworkFee, rollupFee)
 
       // Calculate the max spendable amount which accounts for all fees:
-      const maxSpendable = sub(balance, totalFee)
+      let maxSpendable = sub(balance, totalFeeInitial)
+
+      // For OP-stack chains, refine once or twice by recomputing the L1 fee
+      // using the candidate amount, then subtract a small 1% margin.
+      if (this.optimismRollupParams != null) {
+        // Up to two refinement passes
+        for (let i = 0; i < 2; i++) {
+          const refinedTxData: CalcOptimismRollupFeeParams = {
+            baseFee: this.optimismRollupParams.baseFee,
+            baseFeeScalar: this.optimismRollupParams.baseFeeScalar,
+            blobBaseFee: this.optimismRollupParams.blobBaseFee,
+            blobBaseFeeScalar: this.optimismRollupParams.blobBaseFeeScalar,
+            nonce: this.otherData.unconfirmedNextNonce,
+            gasLimit: miningFees.gasLimit,
+            to: publicAddress,
+            value: decimalToHex(maxSpendable),
+            data,
+            chainParams: this.networkInfo.chainParams
+          }
+          const refinedL1 = calcOptimismRollupFees(refinedTxData)
+          const refinedTotalFee = add(primaryNetworkFee, refinedL1)
+          const refinedAmount = sub(balance, refinedTotalFee)
+          // Stop if converged or increased
+          if (lte(maxSpendable, refinedAmount)) break
+          maxSpendable = refinedAmount
+        }
+
+        // Apply a final 1% margin
+        const margin = ceil(div(mul(maxSpendable, '10'), '1000'), 0)
```

Comment:
Fuzzing using the masSpendable amount can be unnecessarily expensive for large transfers. It would be preferable to base this on what the worst-case L1 fee could be